
<H2><A NAME="SECTION02551000000000000000">
throw y catch son Sentencias de Control</A>
</H2>

<P>
<A NAME="4332"><tex2html_anchor_invisible_mark></A><SPAN  CLASS="textit">throw</SPAN> and <A NAME="4334"><tex2html_anchor_invisible_mark></A><SPAN  CLASS="textit">catch</SPAN> are <A NAME="tex2html197"
  HREF="http://ruby-doc.org/core-2.0.0/Kernel.html"><TT>Kernel</TT></A><#4276#><#4276#> methods that define a control structure
that can be thought of as a multilevel <tex2html_verb_mark>212<tex2html_verb_mark>. 

<P>
<tex2html_verb_mark>213<tex2html_verb_mark> doesn’t just
<tex2html_verb_mark>214<tex2html_verb_mark> out of the current loop or block but can actually transfer
out any number of levels, causing the block defined with a <tex2html_verb_mark>215<tex2html_verb_mark>
to exit. 

<P>
The <tex2html_verb_mark>216<tex2html_verb_mark> need not even be in the same method as the <tex2html_verb_mark>217<tex2html_verb_mark>.
It can be in the calling method, or somewhere even further up the
call stack.

<P>
Languages like Java and JavaScript allow loops to be named or labeled
with an arbitrary prefix. When this is done, a control structure
known as a <A NAME="4339"><tex2html_anchor_invisible_mark></A><SPAN  CLASS="textit">labeled break</SPAN> causes the named loop to exit. 

<P>
Ruby’s
<tex2html_verb_mark>218<tex2html_verb_mark> method defines a labeled block of code, and Ruby’s <tex2html_verb_mark>219<tex2html_verb_mark>
method causes that block to exit. 

<P>
But <tex2html_verb_mark>220<tex2html_verb_mark> and <tex2html_verb_mark>221<tex2html_verb_mark> are much more
general than a labeled <tex2html_verb_mark>222<tex2html_verb_mark>. For one, it can be used with any kind
of statement and is not restricted to loops. 

<P>
More profoundly, a
<tex2html_verb_mark>223<tex2html_verb_mark> can propagate up the call stack to cause a block in an invoking
method to exit.

<P>
If you are familiar with languages like Java and JavaScript, then
you probably recognize <tex2html_verb_mark>224<tex2html_verb_mark> and <tex2html_verb_mark>225<tex2html_verb_mark> as the keywords those languages
use for raising and handling exceptions. 

<P>
Ruby does exceptions
differently, using <tex2html_verb_mark>226<tex2html_verb_mark> and <tex2html_verb_mark>227<tex2html_verb_mark>, which we’ll learn about later
in this chapter. 

<P>
But the parallel to exceptions is intentional.
Calling <tex2html_verb_mark>228<tex2html_verb_mark> is very much like raising an exception. 

<P>
And the way
a <tex2html_verb_mark>229<tex2html_verb_mark> propagates out through the lexical scope and then up the
call stack is very much the same as the way an exception propagates out and up. 

<P>
<#4278#><I>Despite the similarity to exceptions, it is
best to consider</I><#4278#> <tex2html_verb_mark>230<tex2html_verb_mark> <#4279#><I>and</I><#4279#> <tex2html_verb_mark>231<tex2html_verb_mark> <#4280#><I>as a general-purpose (if perhaps
infrequently used) control structure rather than an exception
mechanism</I><#4280#>. 

<P>
If you want to signal an error or exceptional condition,
use <tex2html_verb_mark>232<tex2html_verb_mark> instead of <tex2html_verb_mark>233<tex2html_verb_mark>.

<P>
The following code demonstrates how <tex2html_verb_mark>234<tex2html_verb_mark> and <tex2html_verb_mark>235<tex2html_verb_mark> can be used to <tex2html_verb_mark>236<tex2html_verb_mark> out of nested loops:

<P>
<PRE><tex2html_verbatim_mark>verbatim397#</PRE>
Note that the <tex2html_verb_mark>237<tex2html_verb_mark> method takes a symbol argument and a block. 

<P>
It
executes the block and returns when the block exits or when the
specified symbol is thrown. 

<P>
<tex2html_verb_mark>238<tex2html_verb_mark> also expects a symbol as its
argument and causes the corresponding <tex2html_verb_mark>239<tex2html_verb_mark> invocation to return.

<P>
If no <tex2html_verb_mark>240<tex2html_verb_mark> call matches the symbol passed to <tex2html_verb_mark>241<tex2html_verb_mark>, then a 
<tex2html_verb_mark>242<tex2html_verb_mark>
exception is <tex2html_verb_mark>243<tex2html_verb_mark>d.

<P>
